bc_obps/registration/api/operation.py

```diff
@@ -190,112 +183,117 @@ def create_operation(request, payload: OperationCreateIn):
-                },
-            )
-            al.set_create_or_update(modifier=user)
-            operation.application_lead = al
-
-    # Update other attributes as needed
-    excluded_fields = [
-        "operator",
-        "naics_code",
-        "documents",
-        "contacts",
-        "reporting_activities",
-        "regulated_products",
-        "application_lead",
-    ]
-
+    if payload.operator:
+        operation.operator_id = payload.operator
+
+    if payload.naics_code:
+        operation.naics_code_id = payload.naics_code
+
+    application_lead_address_id = None
+    application_lead_id = payload.application_lead
+    if application_lead_id:
+        application_lead = Contact.objects.get(id=application_lead_id)
+        application_lead_address_id = application_lead.address.id if application_lead.address else None
+
+    is_application_lead_external = payload.is_application_lead_external
+    if is_application_lead_external is True:  # the user has assigned a contact
+        address, _ = Address.objects.update_or_create(
+            id=application_lead_address_id,
+            defaults={
+                "street_address": payload.street_address,
+                "municipality": payload.municipality,
+                "province": payload.province,
+                "postal_code": payload.postal_code,
+            },
+        )
+        eal, _ = Contact.objects.update_or_create(
+            id=application_lead_id,
+            defaults={
+                "first_name": payload.first_name,
+                "last_name": payload.last_name,
+                "position_title": payload.position_title,
+                "email": payload.email,
+                "phone_number": payload.phone_number,
+                "business_role": BusinessRole.objects.get(role_name="Operation Registration Lead"),
+                # "business_role": operation_registration_lead_business_role,
+                "address": address,
+            },
+        )
+        eal.set_create_or_update(modifier=user)
+        operation.application_lead = eal
+
+    if is_application_lead_external is False:  # the lead is the user
+        al, _ = Contact.objects.update_or_create(
+            id=application_lead_id,
+            defaults={
+                "first_name": user.first_name,
+                "last_name": user.last_name,
+                "position_title": user.position_title,
+                "email": user.email,
+                "phone_number": user.phone_number,
+                "business_role": BusinessRole.objects.get(role_name="Operation Registration Lead"),
+                # "business_role": operation_registration_lead_business_role,
+                "address": user.address,
+            },
+        )
+        al.set_create_or_update(modifier=user)
+        operation.application_lead = al
+
+    # updating only a subset of fields (using all fields would overwrite the existing ones)
```

Comment:
The `OperationUpdateIn` schema has many other fields that we handle separately. Therefore, I believe it would be safe to update only those fields that are visible to the user on the form. This approach will prevent unnecessary database queries when the user has not made any changes to the data.

Reply:
Thanks for explaining!